### DISCUSSION TITLE
Porting to Anycubic 4MAX PRO 2.0

### DISCUSSION THREAD

**Discussion by etet100** [bigtreetech/BIGTREETECH-TouchScreenFirmware#2332] [Show and tell]
Hi. I was tired of problems with Anycubic 4max pro 2.0 display (bugs, lack of modern features, weird protocol, no official marlin 2.x support) so i decided to try to port BIGTREETECH-TouchScreenFirmware. The display is already partially working so i'm looking for people interested is such change (alpha testers).

Some important facts:

- ability to program stm32 chip is required
- this change is irreversible (unless someone helps me to dump memory from a working display)
- ANYCUBIC V0.0.2 displays only

And some photos:

![anycubic_v0_0_2_3](https://user-images.githubusercontent.com/22624537/149999549-814649bf-d6f1-48cc-9642-6a3624f64caf.jpg)
![anycubic_v0_0_2_2](https://user-images.githubusercontent.com/22624537/149999565-9f32b0de-6b32-466f-abc6-905152b44cc5.jpg)
![anycubic_v0_0_2_1](https://user-images.githubusercontent.com/22624537/149999567-6e0ca944-0942-437a-8c62-02938fde18f5.jpg)

**Comment by Beta-Ray-Bill** [comment]
Hey ! I have a Anycubic 4max pro 2like you and a was looking for a display firmware for my printer. I found the big tree tech one but I cannot update it via SD, and after that I found your conversation !! I am really interessed by your firmware and by the alpha testing. 

I have two Anycubic TFT 0.0.2 displays, one with OEM firmware and another one bought on ebay dedicated to the modifcations. I have alos prepared a custom Marlin based on OEM one, not base on marlin 2.0 because I'm not sur how to create a new Malin from scratch... 

The second problem is that I don't know how to program stm32 chipset... I have already bought a STlink V2 clone on internet, which seems to be the same as yours, but I don't know how to use it... Could you help me to understand how to use it and to uplad the firmware ?

It will be really cool if you can help me, because right now I am stuck and I cannot go any furhter with the firmware modification. 

Thank you !

**Reply by etet100** [reply]
Hello. I have decided to drop this project due to the lack of interest. And there is more then one type of anycubic v0.0.2. They used to use stm32 cpu but not they use some chinese clone (Nations N32xxx). Second one is even harder to port/flash/debug.

**Comment by Selfmade-RuLeZ** [comment]
Hi, 
I would appreciate if you could give us infos which firmware you have flashed. 

I was able to dump the current firmware of this display with openocd and a raspberry pi 3. I am even able to flash the firmware, the problem for now is just that I don't know what firmware to use. 

If you need the original dump of the stm32f103 I could give it to you, but uploading would be result in a DMCA request I guess.

**Comment by Beta-Ray-Bill** [comment]
Hello,

I didn't understand how to dump the OEM firmware ans how to load the new one, so I didn't go further... I was looking on a STLink V2 clone on Amazon but I couldn't find the software and how to plug the screen...

How did you do to dump the firmware ? How do you plug the screen on the Raspberry Pi and which software did you use ? Can you give me all the procedure you apply, because I don't know how to procede... I think I can try and report to you which firmware is the best.

**Reply by Selfmade-RuLeZ** [reply]
Prerequisites are: 
- 4 Pin Header soldererd to the TFT module
- 5 Jumper Cables f/f 
- working openocd installation on a Raspberry Pi 

Caution:  
I tested it on a Raspberry Pi 3 and with the STM32F103VC microcontroller, maybe you have adopt something to match other platforms. 
3.3V is needed to pull up BOOT0 i guess, i did not tested it without 3.3V line. 5V on the main connector is needed to hold the current. 

Steps:  
Connect the TFT to the Pi: 

|  Pi 	| TFT  	|
|---	|---	|
| 5V  	|  5V 	|
| 3.3V  	| VDD  	|
| GND  	| GND  	|
| GPIO 24  	| SW 	|
| GPIO 25  	| CLK  	|

Create a config file `rpi-interface.cfg` for your raspberry pi interface: 

```
adapter driver bcm2835gpio
bcm2835gpio peripheral_base 0x3F000000
bcm2835gpio speed_coeffs 146203 36
bcm2835gpio swd_nums 25 24
reset_config none
```

To dump the current firmware you have to use following command: 

```
sudo openocd -f ./rpi-interface.cfg -c "transport select swd" -f target/stm32f1x.cfg -c init -c "reset halt" -c "flash read_bank 0 firmware_dump.bin 0 0x40000" -c "reset" -c shutdown
```

to write a firmware: 

```
sudo openocd -f ./rpi-interface.cfg -c "transport select swd" -f target/stm32f1x.cfg -c init -c "reset halt" -c "flash write_image erase firmware_dump.bin 0x08000000" -c "reset" -c shutdown
```

I hope i can help you with this "tutorial".

**Comment by Beta-Ray-Bill** [comment]
Oh  what a dumb I am. I didn't see the 4 pin to be soldered... I already have a raspberry pi and a STM32F103VC controller, so I just need to solder the header and try the installation.
Which OS did you use on the Pi ? Raspbian ? Actually, I have octopi running on mine, but I have another sd card if needed.
I think I will need the original dump of your stm32f103, because I think my original TFT firmware is broken (nothing happen when I plug my TFT to the printer and I turn it on), so do you think you could create a temporary Google Drive link ? Just the time I need to download it or maybe you can use wetranfer ?

**Reply by Selfmade-RuLeZ** [reply]
I found out, that the first 64kB are the bootloader. After the bootloader is the flash image starting at 0x01000 with 192 kB. 
Try to put this on the root of a sd card  (formatted in FAT32, if there are problems use one with at max. 8GB and 4K cluster or create a partition with max 8GB) and stick it into the TFT. Afterwards powercycle it (maybe connected to the printer). 

Files with the ending `x.bin` (e.g. `firmware.x.bin`) will get loaded into the flash by the bootloader. 

First try this with the sdcard method: https://www.mediafire.com/file/olskacfl8t0m48h/firmware.x.bin/file

If this does not work, you have to connect your pi and flash the file attached below over openocd. 

https://www.mediafire.com/file/3kdgyd3d99699qm/firmware_full.bin/file

Edit: 

I flashed a random image from this repo here to "brick" the TFT and recovered it by putting the `firmware.x.bin` file on the sdcard and powercycling the TFT. So if you just want to recover your TFT here, you are good to go i guess. 

For others: If @etet100 could give us infos which image he used, we could try to develop a on firmware based on the image.

**Reply by etet100** [reply]
You are great! I wonder if it also works with n32 chip.

**Reply by Selfmade-RuLeZ** [reply]
I would not try it before you dumped the flash somehow 😁 

Edit: I saw in the polish forums (Google Translate FTW!), that you managed to dump the file. You could analyze it visually und see where the padding of the bootloader is. Afterwards delete the bootloder with the padding from the file and viola, you have the sdcard flashable file. No warranty for anything 😁

**Reply by etet100** [reply]
@Selfmade-RuLeZ  are you a software developer?

**Reply by Selfmade-RuLeZ** [reply]
Junior developer in the last semester of softwareengineering bachelor course.

**Reply by etet100** [reply]
I was just wondering if you are willing/able to try update/finish my port.

**Reply by Selfmade-RuLeZ** [reply]
I guess I have not enough time currently, BUT I'm very interested in your first try of the port. Maybe we can improve it together step by step with the community :)

**Reply by etet100** [reply]
I don't like communities. Too many people to argue with.

**Reply by etet100** [reply]
@Selfmade-RuLeZ Hi. Could you upload both firmware.bin files once again please? Both urls are outdated.

**Reply by mmaciag1991** [reply]
Hey do you have this firmware?

**Reply by etet100** [reply]
no

**Reply by Selfmade-RuLeZ** [reply]
I totally forgot about this one, my bad. I'm going to take a look on my Computer later this day if I still have the right files. They should be there if I didn't mess things up.

**Reply by mmaciag1991** [reply]
Cool. I found this thread because I flashed this tft with the wrong software and now I have neither the original software nor any alternative. I'm a C# programmer, I haven't had much contact with microcontrollers, but my cpp stm colleagues from work will help me.

**Reply by mmaciag1991** [reply]
![IMG_20230917_161829.jpg](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/assets/46632727/4a2080d5-cee3-423e-b5e9-d85d72c105d7)

**Reply by mmaciag1991** [reply]
I have the same display but with a different processor, it's probably some stm clone

**Reply by Selfmade-RuLeZ** [reply]
https://www.mediafire.com/file/h3mx72s7uazqfo4/firmware.zip/file

Again, 14 days. Feel free to store it somewhere you like. I hope it will work with this processor, the firmware could do, but the bootloader could be a dealbreaker here. Again, there are both files included, the firmware.x.bin (without bootloader) and the full bootloader. Flash instructions some comments above.

**Reply by mmaciag1991** [reply]
Thank you very mutch :)

**Reply by jaksonqwear** [reply]
Hi, I'm having the same problem. I have the same display, but I put a bad "original" firmware on it. After that it won't turn on at all. When connecting a flash drive with any firmware, nothing happens. I don't have a pi but I have a USB-STL-ISP, can you tell me if I can use it for this task, I'm not very good at it.

**Comment by GizmoIIID** [comment]
I have an AnyCubic v0.0.2 display and a 4max Pro 3d printer. I am ready to participate in the trials if necessary.

**Comment by Sturmxk** [comment]
Hello, does anyone have the original firmware for me?
